gdal/frmts/jpegxr/jpegxrdataset.cpp

```diff
@@ -0,0 +1,1202 @@
+    {
+        if (!pImageEncode)
+            throw JPEGXRException(WMP_errNotInitialized);
+
+        U32 const cLine = static_cast<U32>(nYSize);
+        U32 const cbStride = static_cast<U32>(nXStride);
+
+        ERR nErr = pImageEncode->WritePixels(pImageEncode,
+                                             cLine,
+                                             pabyData,
+                                             cbStride);
+        if (nErr != WMP_errSuccess)
+            throw JPEGXRException(nErr);
+    }
+
+private:
+    PKFactory* pFactory;
+    PKCodecFactory* pCodecFactory;
+    PKImageEncode* pImageEncode;
+    WMPStream* pEncodeStream;
+};
+
+
+/************************************************************************/
+/* ==================================================================== */
+/*                           JPEGXRDataset                              */
+/* ==================================================================== */
+/************************************************************************/
+
+class JPEGXRDataset final : public GDALPamDataset
+{
+    friend class JPEGXRRasterBand;
+private:
+    JPEGXRDecoder oDecoder;
+    GByte* pabyUncompressedData;
+    bool bHasUncompressed;
+
+    CPLErr Uncompress();
+
+public:
+    JPEGXRDataset();
+    ~JPEGXRDataset();
+
+    static int          Identify(GDALOpenInfo*);
+    static GDALDataset* Open(GDALOpenInfo*);
+    static GDALDataset* CreateCopy(char const* pszFilename,
+                                   GDALDataset* poSrcDS,
+                                   int bStrict,
+                                   char** papszOptions,
+                                   GDALProgressFunc pfnProgress,
+                                   void* pProgressData);
+};
+
+/************************************************************************/
+/* ==================================================================== */
+/*                         JPEGXRRasterBand                             */
+/* ==================================================================== */
+/************************************************************************/
+
+class JPEGXRRasterBand final : public GDALPamRasterBand
+{
+    friend class JPEGXRDataset;
+
+  public:
+
+    JPEGXRRasterBand(JPEGXRDataset* poDS, int nBand);
+
+    CPLErr          IReadBlock(int, int, void*) override;
+    GDALColorInterp GetColorInterpretation() override;
+};
+
+
+/************************************************************************/
+/*                        JPEGXRRasterBand()                            */
+/************************************************************************/
+
+JPEGXRRasterBand::JPEGXRRasterBand(JPEGXRDataset* poDSIn, int nBandIn)
+{
+    poDS = poDSIn;
+    nBand = nBandIn;
+    nRasterXSize = poDSIn->nRasterXSize;
+    nRasterYSize = poDSIn->nRasterYSize;
+    nBlockXSize = nRasterXSize;
+    nBlockYSize = nRasterYSize;
```

Comment:
Exposing the image as a single block is rather "rude" (if it is big) even if it is entirely decompressed. I'd suggest exposing nBlockYSize = 1 (and dealing with that in IReadBlock())